src/xrpld/app/tx/detail/DepositPreauth.cpp

```diff
@@ -149,35 +221,98 @@ DepositPreauth::doApply()
+        // check the starting balance because we want to allow dipping into the
+        // reserve to pay fees.
+        {
+            STAmount const reserve{view().fees().accountReserve(
+                sleOwner->getFieldU32(sfOwnerCount) + 1)};
+
+            if (mPriorBalance < reserve)
+                return tecINSUFFICIENT_RESERVE;
+        }
+
+        // Preclaim already verified that the Preauth entry does not yet exist.
+        // Create and populate the Preauth entry.
+
+        STArray const& authCred(ctx_.tx.getFieldArray(sfAuthorizeCredentials));
+        Keylet const preauthKey = keylet::depositPreauth(account_, authCred);
+        auto slePreauth = std::make_shared<SLE>(preauthKey);
+        if (!slePreauth)
+            return tefINTERNAL;
+
+        slePreauth->setAccountID(sfAccount, account_);
+        STArray& credentialsArray =
+            slePreauth->peekFieldArray(sfAuthorizeCredentials);
+        credentialsArray.reserve(authCred.size());
+
+        // eliminate duplicates
+        std::unordered_set<uint256> hashes;
+        for (auto const& o : authCred)
+        {
+            auto [it, ins] = hashes.insert(sha512Half(
+                o.getAccountID(sfIssuer), o.getFieldVL(sfCredentialType)));
+            if (ins)
+            {
+                credentialsArray.push_back(o);
+            }
+            else
+            {
+                JLOG(j_.trace())
+                    << "DepositPreauth duplicate removed: " << o.getText();
+            }
+        }
+
+        view().insert(slePreauth);
+
+        auto const page = view().dirInsert(
+            keylet::ownerDir(account_), preauthKey, describeOwnerDir(account_));
+
+        JLOG(j_.trace()) << "Adding DepositPreauth to owner directory "
+                         << to_string(preauthKey.key) << ": "
+                         << (page ? "success" : "failure");
+
+        if (!page)
+            return tecDIR_FULL;
+
+        slePreauth->setFieldU64(sfOwnerNode, *page);
+
+        // If we succeeded, the new entry counts against the creator's reserve.
+        adjustOwnerCount(view(), sleOwner, 1, j_);
+    }
+    else if (ctx_.tx.isFieldPresent(sfUnauthorizeCredentials))
+    {
+        auto const preauthKey = keylet::depositPreauth(
+            account_, ctx_.tx.getFieldArray(sfUnauthorizeCredentials));
+        return DepositPreauth::removeFromLedger(view(), preauthKey.key, j_);
+    }
+
     return tesSUCCESS;
 }
 
 TER
 DepositPreauth::removeFromLedger(
-    Application& app,
     ApplyView& view,
     uint256 const& preauthIndex,
     beast::Journal j)
 {
-    // Verify that the Preauth entry they asked to remove is
-    // in the ledger.
-    std::shared_ptr<SLE> const slePreauth{
-        view.peek(keylet::depositPreauth(preauthIndex))};
-    if (!slePreauth)
-    {
-        JLOG(j.warn()) << "Selected DepositPreauth does not exist.";
-        return tecNO_ENTRY;
-    }
```

Comment:
This still feels like a transaction-breaking change.

Reply:
fixed